pkg/bgp/speaker/speaker.go

```diff
@@ -0,0 +1,217 @@
+}
+
+// OnUpdateService notifies the Speaker of an update to a service.
+func (s *Speaker) OnUpdateService(name string, svc *slim_corev1.Service) {
+	fmt.Printf("chris OnUpdateService(%v)\n", name)
+
+	svcID := k8s.ParseServiceID(svc)
+
+	eps := new(metallbspr.Endpoints)
+	epsFromSvc := s.endpointsGetter.GetEndpointsOfService(svcID)
+	if epsFromSvc != nil {
+		eps = convertInternalEndpoints(epsFromSvc)
+	}
+
+	s.Lock()
+	s.services[svcID] = svc
+	s.Unlock()
+
+	s.queue.Add(epEvent{
+		name: svc.GetName(),
+		svc:  convertService(svc),
+		eps:  eps,
+	})
+}
+
+// OnDeleteService notifies the Speaker of a delete of a service.
+func (s *Speaker) OnDeleteService(name string, svc *slim_corev1.Service) {
+	fmt.Printf("chris OnDeleteService(%v)\n", name)
+
+	s.Lock()
+	delete(s.services, k8s.ParseServiceID(svc))
+	s.Unlock()
+
+	// Passing nil as the service will force the MetalLB speaker to withdraw
+	// the BGP announcement.
+	s.queue.Add(svcEvent{
+		name: svc.GetName(),
+		svc:  nil,
+		eps:  nil,
+	})
+}
+
+// OnUpdateEndpoints notifies the Speaker of an update to the backends of a
+// service.
+func (s *Speaker) OnUpdateEndpoints(eps *slim_corev1.Endpoints) {
+	svcID := k8s.ParseEndpointsID(eps)
+
+	s.Lock()
+	defer s.Unlock()
+	if svc, ok := s.services[svcID]; ok {
+		s.queue.Add(epEvent{
+			name: svc.GetName(),
+			svc:  convertService(svc),
+			eps:  convertEndpoints(eps),
+		})
+	}
+}
+
+// OnUpdateNode notifies the Speaker of an update to a node.
+func (s *Speaker) OnUpdateNode(node *slim_corev1.Node) {
+	fmt.Printf("chris OnUpdateNode(%v)\n", node.GetName())
+	s.queue.Add(nodeEvent(&node.Labels))
+}
+
+// RegisterSvcCache registers the K8s watcher cache with this Speaker.
+func (s *Speaker) RegisterSvcCache(cache endpointsGetter) {
+	s.endpointsGetter = cache
+}
+
+// endpointsGetter abstracts the github.com/cilium/cilium/pkg/k8s.ServiceCache
+// object. The cache holds all services and endpoints (backends) from the K8s
+// watchers.
+type endpointsGetter interface {
+	GetEndpointsOfService(svcID k8s.ServiceID) *k8s.Endpoints
+}
+
+func convertService(in *slim_corev1.Service) *metallbspr.Service {
+	if in == nil {
+		return nil
+	}
+	ing := make([]v1.LoadBalancerIngress, len(in.Status.LoadBalancer.Ingress))
+	for i := range in.Status.LoadBalancer.Ingress {
+		ing[i].IP = in.Status.LoadBalancer.Ingress[i].IP
+	}
+	fmt.Printf("chris convertService(%v) ingress=%v\n", in.GetName(), ing)
+	return &metallbspr.Service{
+		Type:          string(in.Spec.Type),
+		TrafficPolicy: string(in.Spec.ExternalTrafficPolicy),
+		Ingress:       ing,
+	}
+}
+
+func convertInternalEndpoints(in *k8s.Endpoints) *metallbspr.Endpoints {
+	if in == nil {
+		return nil
+	}
+	out := new(metallbspr.Endpoints)
+	for ip, be := range in.Backends {
+		ep := metallbspr.Endpoint{
+			IP:       ip,
+			NodeName: &be.NodeName,
+		}
+		if be.Ready {
+			out.Ready = append(out.Ready, ep)
+		}
+		if !be.Ready {
+			out.NotReady = append(out.NotReady, ep)
```

Comment:
One of my questions while reading the commit message was: will the BGP announcement be withdrawn if the status of a  backend changes from healthy to unhealthy? Am i correct to assume that this happens here?

Reply:
Yes, it will. I thought I mentioned that in the commit msg, but apparently not:

> For Endpoints updates, we watch for these because we must know when a
> service has healthy backends (endpoints). Only when that happens can a
> BGP announcement be made. A mapping is stored internally for services to
> endpoints, so that we may cross-reference them back to a Service. For
> clarity, an announcement is never made unless a service has healthy
> backends.

I will add it in that parapraph and clarify it a bit because reading now, it doesn't flow as nicely as I intend :). Thanks!